src/cmdline/mod.rs

```diff
@@ -116,134 +97,98 @@ impl Cmdline {
     /// let cl = Cmdline::new(100);
     /// ```
     /// [`Cmdline`]: struct.Cmdline.html
-    pub fn new(capacity: usize) -> Cmdline {
-        assert_ne!(capacity, 0);
+    pub fn new(cmdline_capacity: usize) -> Cmdline {
+        assert!(cmdline_capacity > CMDLINE_NO_OF_AUXILIARY_CHARS);
+
         Cmdline {
-            line: String::with_capacity(capacity),
-            capacity,
+            boot_args: String::with_capacity(cmdline_capacity),
+            init_args: String::with_capacity(cmdline_capacity),
+            cmdline_capacity,
         }
     }
 
     fn has_capacity(&self, more: usize) -> Result<()> {
-        let needs_space = if self.line.is_empty() { 0 } else { 1 };
-        if self.line.len() + more + needs_space < self.capacity {
+        if self.boot_args.len() + self.init_args.len() + more + CMDLINE_NO_OF_AUXILIARY_CHARS
+            <= self.cmdline_capacity
+        {
             Ok(())
         } else {
             Err(Error::TooLarge)
         }
     }
 
-    fn start_push(&mut self) {
-        if !self.line.is_empty() {
-            self.line.push(' ');
-        }
-    }
-
-    fn end_push(&mut self) {
-        // This assert is always true because of the `has_capacity` check that each insert method
-        // uses.
-        assert!(self.line.len() < self.capacity);
-    }
-
-    /// Validates and inserts a key-value pair into this command line.
+    /// Validates and inserts a string representing a boot arg to the current command line.
     ///
     /// # Arguments
     ///
-    /// * `key` - Key to be inserted in the command line string.
-    /// * `val` - Value corresponding to `key`.
+    /// * `slug` - String representing a boot arg that needs to be added to the cmdline.
     ///
     /// # Examples
     ///
     /// ```rust
     /// # use linux_loader::cmdline::*;
     /// # use std::ffi::CString;
     /// let mut cl = Cmdline::new(100);
-    /// cl.insert("foo", "bar");
+    /// cl.insert_boot_arg("foobar").expect("");
     /// let cl_cstring = CString::new(cl).unwrap();
-    /// assert_eq!(cl_cstring.to_str().unwrap(), "foo=bar");
+    /// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
     /// ```
-    pub fn insert<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()> {
```

Comment:
Ok, that was one aspect I was not sure about. I saw that we don't use it anywhere in Firecracker so I decided to remove it. However, from the abstraction point of view, yes... you are right... it would be better to keep this feature available. I'll add the methods related to key/value back into the commit.

Reply:
This is not only used by Firecracker, but other customers as well. We shouldn't remove existing functionality.